packages/ERTP/src/mathHelpers/strSetMathHelpers.js

```diff
@@ -25,42 +30,98 @@ const strSetMathHelpers = harden({
   doCoerce: list => {
     assert(passStyleOf(list) === 'copyArray', 'value must be an array');
     list.forEach(elem => assert.typeof(elem, 'string'));
-    checkForDupes(list);
-    return list;
+    assertUniqueSorted(list);
+    return harden(list);
   },
   doGetEmpty: _ => identity,
   doIsEmpty: list => passStyleOf(list) === 'copyArray' && list.length === 0,
   doIsGTE: (left, right) => {
-    const leftSet = new Set(left);
-    const leftHas = elem => leftSet.has(elem);
-    return right.every(leftHas);
+    let leftI = 0;
+    let rightI = 0;
+    const leftLen = left.length;
+    const rightLen = right.length;
+    while (leftI < leftLen && rightI < rightLen) {
+      const leftStr = left[leftI];
+      const rightStr = right[rightI];
+      if (leftStr < rightStr) {
+        // an element of left not in right. Fine
+        leftI += 1;
+      } else if (leftStr > rightStr) {
+        // an element of right not in left.
+        return false;
+      } else {
+        leftI += 1;
+        rightI += 1;
+      }
+    }
+    // Are there no elements of right remaining?
+    return rightI >= rightLen;
   },
   doIsEqual: (left, right) => {
-    const leftSet = new Set(left);
-    const leftHas = elem => leftSet.has(elem);
-    return left.length === right.length && right.every(leftHas);
+    if (left.length !== right.length) {
+      return false;
+    }
+    return left.every((leftStr, i) => leftStr === right[i]);
   },
   doAdd: (left, right) => {
-    const union = new Set(left);
-    const addToUnion = elem => {
+    const result = [];
+    let leftI = 0;
+    let rightI = 0;
+    const leftLen = left.length;
+    const rightLen = right.length;
+    while (leftI < leftLen && rightI < rightLen) {
+      const leftStr = left[leftI];
+      const rightStr = right[rightI];
       assert(
-        !union.has(elem),
-        details`left and right have same element ${elem}`,
+        leftStr !== rightStr,
+        details`left and right have same element ${leftStr}`,
       );
-      union.add(elem);
-    };
-    right.forEach(addToUnion);
-    return harden(Array.from(union));
+      if (leftStr < rightStr) {
+        result.push(leftStr);
+        leftI += 1;
+      } else {
+        result.push(rightStr);
+        rightI += 1;
+      }
+    }
+    if (leftI < leftLen) {
+      result.push(left[leftI]);
+    } else if (rightI < rightLen) {
+      result.push(right[rightI]);
+    }
```

Comment:
I think these leftover processing loops should be while loops. Also the two in `doSubtract` below.

Reply:
Hmm. not sure. need closer examination than I'm about to give it.